File: app/src/main/java/dev/ragnarok/fenrir/fragment/search/criteria/AudioSearchCriteria.java
package dev.ragnarok.fenrir.fragment.search.criteria;

import android.os.Parcel;

import androidx.annotation.NonNull;

import java.util.ArrayList;

import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.fragment.search.options.SimpleBooleanOption;
import dev.ragnarok.fenrir.fragment.search.options.SpinnerOption;

public class AudioSearchCriteria extends BaseSearchCriteria {
    public static final int KEY_SEARCH_ADDED = 1;
    public static final int KEY_SEARCH_BY_ARTIST = 2;
    public static final int KEY_SEARCH_AUTOCOMPLETE = 3;
    public static final int KEY_SEARCH_WITH_LYRICS = 4;
    public static final int KEY_SORT = 5;
    public static final Creator<AudioSearchCriteria> CREATOR = new Creator<AudioSearchCriteria>() {
        @Override
        public AudioSearchCriteria createFromParcel(Parcel in) {
            return new AudioSearchCriteria(in);
        }

        @Override
        public AudioSearchCriteria[] newArray(int size) {
            return new AudioSearchCriteria[size];
        }
    };

    public AudioSearchCriteria(String query, boolean by_artist, boolean in_main_page) {
        super(query);
        SpinnerOption sort = new SpinnerOption(KEY_SORT, R.string.sorting, true);
        sort.available = new ArrayList<>(3);
        sort.available.add(new SpinnerOption.Entry(0, R.string.by_date_added));
        sort.available.add(new SpinnerOption.Entry(1, R.string.by_relevance));
        sort.available.add(new SpinnerOption.Entry(2, R.string.by_duration));
        appendOption(sort);

        appendOption(new SimpleBooleanOption(KEY_SEARCH_ADDED, R.string.my_saved, true, in_main_page));
        appendOption(new SimpleBooleanOption(KEY_SEARCH_BY_ARTIST, R.string.by_artist, true, by_artist));
        appendOption(new SimpleBooleanOption(KEY_SEARCH_AUTOCOMPLETE, R.string.auto_compete, true));
        appendOption(new SimpleBooleanOption(KEY_SEARCH_WITH_LYRICS, R.string.with_lyrics, true));
    }

    protected AudioSearchCriteria(Parcel in) {
        super(in);
    }

    @NonNull
    @Override
    public AudioSearchCriteria clone() throws CloneNotSupportedException {
        return (AudioSearchCriteria) super.clone();
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/fragment/search/criteria/AudioSearchCriteria.kt
package dev.ragnarok.fenrir.fragment.search.criteria

import android.os.Parcel
import android.os.Parcelable
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.fragment.search.options.SimpleBooleanOption
import dev.ragnarok.fenrir.fragment.search.options.SpinnerOption

class AudioSearchCriteria : BaseSearchCriteria {
    constructor(query: String?, by_artist: Boolean, in_main_page: Boolean) : super(query) {
        val sort = SpinnerOption(KEY_SORT, R.string.sorting, true)
        sort.available = ArrayList(3)
        sort.available.add(SpinnerOption.Entry(0, R.string.by_date_added))
        sort.available.add(SpinnerOption.Entry(1, R.string.by_relevance))
        sort.available.add(SpinnerOption.Entry(2, R.string.by_duration))
        appendOption(sort)
        appendOption(SimpleBooleanOption(KEY_SEARCH_ADDED, R.string.my_saved, true, in_main_page))
        appendOption(SimpleBooleanOption(KEY_SEARCH_BY_ARTIST, R.string.by_artist, true, by_artist))
        appendOption(SimpleBooleanOption(KEY_SEARCH_AUTOCOMPLETE, R.string.auto_compete, true))
        appendOption(SimpleBooleanOption(KEY_SEARCH_WITH_LYRICS, R.string.with_lyrics, true))
    }

    internal constructor(`in`: Parcel) : super(`in`)

    @Throws(CloneNotSupportedException::class)
    public override fun clone(): AudioSearchCriteria {
        return super.clone() as AudioSearchCriteria
    }

    companion object {
        const val KEY_SEARCH_ADDED = 1
        const val KEY_SEARCH_BY_ARTIST = 2
        const val KEY_SEARCH_AUTOCOMPLETE = 3
        const val KEY_SEARCH_WITH_LYRICS = 4
        const val KEY_SORT = 5

        @JvmField
        val CREATOR: Parcelable.Creator<AudioSearchCriteria> =
            object : Parcelable.Creator<AudioSearchCriteria> {
                override fun createFromParcel(`in`: Parcel): AudioSearchCriteria {
                    return AudioSearchCriteria(`in`)
                }

                override fun newArray(size: Int): Array<AudioSearchCriteria?> {
                    return arrayOfNulls(size)
                }
            }
    }
}